### DISCUSSION TITLE
Allow users to set primary component metadata as part of the generation

### DISCUSSION THREAD

**Discussion by vpetersson** [aquasecurity/trivy#7872] [Ideas]
### Description

I'm part of the CISA Tiger Team for [SBOM Reference Implementation](https://github.com/CISA-SBOM-Community/SBOM-Generation/). We've adopted Trivy as part of our tool chain for the first iteration of our reference implementation.

While Trivy works well in general, we have some feedback that would make it easier to generate NTIA Minimum Element compliant SBOMs (which is the goal of the Tiger Team).

The feedback is as follows:

* Allow users to specify a the following as command line arguments:
  * [Version](https://cyclonedx.org/docs/1.6/json/#metadata_component_version)
  * [Name](https://cyclonedx.org/docs/1.6/json/#metadata_component_name)
  * [License](https://cyclonedx.org/docs/1.6/json/#metadata_component_licenses)
  * [Manufacturer](https://cyclonedx.org/docs/1.6/json/#metadata_component_manufacturer)
  * [Supplier](https://cyclonedx.org/docs/1.6/json/#metadata_component_supplier) 

(Note: Using CycloneDX just as an example)
 
We've worked around these constraints using [sbomasm](https://github.com/interlynk-io/sbomasm) ([example](https://github.com/CISA-SBOM-Community/SBOM-Generation/blob/main/.github/workflows/phase_1_python.yml#L204)), but it would streamline things a lot of this could be provided as part of the generation phase.

### Target

Filesystem

### Scanner

None

**Comment by knqyf263** [comment]
>We've adopted Trivy as part of our tool chain for the first iteration of our reference implementation.

We're glad to hear that! Thanks for sharing.

>Allow users to specify a the following as command line arguments:

While this suggestion makes sense, it will add too many CLI flags and is not flexible. If users have another need, like specifying `components[].purl,` we need to add a new flag.

So, I'd suggest using [JSON Patch](https://datatracker.ietf.org/doc/html/rfc6902) or [JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7386). Users create a JSON patch and pass it to Trivy when generating SBOM, like `trivy image --format cyclonedx --sbom-json-patch ./my-sbom-template.patch my-image:latest`. It wouldn't work with SPDX text format, but I think it's acceptable. I personally prefer JSON Merge Patch, but need to check if it meets the requirements. I'm just thinking out loud. Any feedback is welcome.

**Reply by vpetersson** [reply]
Just to be clear, we're not talking about adding elements for *components*, just the top level component.

It's good to know that there's support for JSON patching in Trivy. We initially used `jq` for patching the JSON, but it's really cumbersome when you need to deliver both SPDX and CycloneDX, which is the deliverable for our working group. 

`sbomasm` does the job for now (in a format agnostic way), but it's our recommendation from the working group is that this really should be a responsibility of the generation tool.

**Reply by knqyf263** [reply]
>Just to be clear, we're not talking about adding elements for components, just the top level component.

Yes, I understand. I meant needs will continue to expand. It may be just the top-level component at the moment, but users may want to specify other metadata, and it is not practical to add a new flag each time they do so.

>It's good to know that there's support for JSON patching in Trivy.

I'm talking about a feature enhancement. It's not supported in Trivy now.

>We initially used jq for patching the JSON, but it's really cumbersome when you need to deliver both SPDX and CycloneDX, which is the deliverable for our working group.

If you need a format-agnostic approach, we need to consider something else because the idea I suggested requires JSON patches for SPDX and CycloneDX, respectively.

What if creating a new [plugin](https://aquasecurity.github.io/trivy/v0.57/docs/plugin/) for `sbomasm`? [Here](https://aquasecurity.github.io/trivy-plugin-index/) is the plugin list. People can call `sbomasm` as part of Trivy, like `trivy sbomasm edit --subject primary-component --name "cool-app" --version "v1.0.0" --type "application" --output cool-app-mod.spdx.json cool-app.spdx.json`. It's just a wrapper of `sbomasm`, but it looks more integrated into Trivy and may provide a better user experience.

Another idea is that since Trivy has [an intermediate representation of SBOM](https://github.com/aquasecurity/trivy/pull/6240), it might be possible to patch against it and convert it to SPDX or CycloneDX format independently.

**Reply by vpetersson** [reply]
I completely understand that this might be a slippery slope.

The plugin approach might be the cleanest solution. I like that. Let me ponder on this for a bit. Maybe a simple Bash plugin would suffice.

**Reply by knqyf263** [reply]
The developer documentation is here.
https://aquasecurity.github.io/trivy/v0.57/docs/plugin/developer-guide/

Some plugins are written in a shell script.
- https://github.com/aquasecurity/trivy-plugin-kubectl/blob/main/trivy-kubectl
- https://github.com/fatihtokus/scan2html/blob/main/scan2html.sh

**Reply by knqyf263** [reply]
If you agree on the plugin approach, we can help with that.

**Reply by knqyf263** [reply]
FYI: Trivy also supports [output plugins](https://aquasecurity.github.io/trivy/v0.57/docs/plugin/user-guide/#output-mode-support). It provides a more integrated UX, like 
```
$ trivy image --format cyconedx --output plugin=sbomasm --output-plugin-arg "--subject primary-component --name "cool-app" --version "v1.0.0" --type "application" --output cool-app-mod.spdx.json myimage:latest"
```

This is the [reference implementation](https://github.com/aquasecurity/trivy-plugin-count).
